### PR TITLE
Bctw 935 add critter form

### PR DESCRIPTION
**Pull request by MacDeluca**
## Pre Requirements

- Must add the new env variable for Critterbase's API host to .env
- CB_API_HOST=https://moe-critterbase-api-dev.apps.silver.devops.gov.bc.ca/
- Note: Will extend the docker make file in the future to generate this value.

## Links to Jira Tickets

- PR using previous BCTW ticket title
- [https://apps.nrs.gov.bc.ca/int/jira/secure/RapidBoard.jspa?rapidView=8685&projectKey=SIMSBIOHUB&view=planning&selectedIssue=SIMSBIOHUB-241&issueLimit=100](url)

## Description of Changes

- This PR contains the new Create Critter (Add Individuals) form.
- Handles general animal details, captures, mortalities, measurements, markings, families
- Inline validation using formik + additional critterbase requests to check if certain values are valid.
- CbSelect component handles retrieving and displaying the critterbase related lookup values.
- Some additional wrapper components / custom tooltips.
- Storing all animal form related strings in i18n.ts
- Mostly net-new components, with some additional refactoring of some duplicated components into a single component used in survey section.

- Note: Does not currently support images or telemetry devices. Will make seperate tickets for those additional features. 

## Testing Notes

- This PR contains changes from both myself (Mac) and Graham.
- Currently this form is sending requests directly to Critterbase. In upcoming tickets this will change to using the SIM's API as a proxy.
- Added a development button for displaying the form values / critterbase payload at the bottom of the form.
- No UI tests written for this yet, waiting till we finalize the UI and talk standards for what to test in the UI etc.
- No critters will display in the survey section after completion, waiting on additional mockups for what that looks like.

### CODE REVIEW

File: api/src/app.ts [hunk 1]
```diff
@@ -128,7 +128,7 @@ try {
  * @param {NextFunction} next
  */
 function validateAllResponses(req: Request, res: Response, next: NextFunction) {
-  const isStrictValidation = !!req['apiDoc']['x-express-openapi-validation-strict'] || false;
+  const isStrictValidation = !!req['apiDoc']?.['x-express-openapi-validation-strict'] || false;
```
Comment (NickPhura) [inline api/src/app.ts]: nit: probably not needed since there are no other api changes

File: app/src/components/dialog/EditDialog.tsx [hunk 2]
```diff
@@ -23,7 +23,7 @@ export interface IEditDialogProps<T> {
    * @type {string}
    * @memberof IEditDialogProps
    */
-  dialogTitle: string;
+  dialogTitle: string | JSX.Element;
```
Comment (NickPhura) [inline app/src/components/dialog/EditDialog.tsx:26]: nit: can you update the jsdoc

There are a few places where the jsdoc could be tweaked (just the formatting, not the descriptions or anything. If you want, this extension is super helpful for jsdoc, I use it constantly: https://marketplace.visualstudio.com/items?itemName=oouo-diogo-perdigao.docthis

File: app/src/contexts/configContext.tsx [hunk 3]
```diff
@@ -42,12 +44,17 @@ const getLocalConfig = (): IConfig => {
...
+  const CB_API_PORT = process.env.REACT_APP_CRITTERBASE_API_PORT;
+
   const API_URL = (API_PORT && `${API_HOST}:${API_PORT}`) || API_HOST || 'localhost';
+  const CB_API_URL = (CB_API_PORT && `${CB_API_HOST}:${CB_API_PORT}`) || CB_API_HOST || 'localhost';
 
   const OBJECT_STORE_URL = process.env.OBJECT_STORE_URL || 'nrs.objectstore.gov.bc.ca';
   const OBJECT_STORE_BUCKET_NAME = process.env.OBJECT_STORE_BUCKET_NAME || 'gblhvt';
   return {
     API_HOST: ensureProtocol(API_URL, 'http://'),
+    CB_API_HOST: CB_API_URL,
```
Comment (NickPhura) [inline app/src/contexts/configContext.tsx]: Can we wrap this in ensureProtocol as well, just to keep it consistent

File: app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+          }}
+          label="Measured Date"
+          name={getAnimalFieldName<IAnimalMeasurement>(NAME, 'measured_timestamp', index)}
+        />
+      </Grid>
+      <Grid item xs={12}>
+        <TextInputToggle label={SurveyAnimalsI18N.animalSectionComment('Measurement')}>
+          <CustomTextField
+            other={{ size: 'small', required: isRequiredInSchema(AnimalMeasurementSchema, 'measurement_comment') }}
+            label="Measurment Comment"
```
Comment (NickPhura) [inline app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx]: nit: spelling

File: app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    if (val < min) {
+      return `Measurement must be greater than ${min}${unit}`;
+    }
+    if (max && val > max) {
+      return `Measurement must be less than ${max}${unit}`;
+    }
+  };
+
+  return (
+    <Fragment key={`meausurement-inputs-${index}`}>
```
Comment (NickPhura) [inline app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx]: nit: spelling

**Comment by NickPhura** [comment]
@MacQSL 

There are a handful of warnings in the app log, mostly dependency array related, a few others. 
(Ignore all of the react-keycloak warnings, those are unrelate and addressing them is still TBD)

Getting some warnings in the form for a bunch of the array sections.
`Encountered two children with the same key, '[object Object]'.`

Also noticing some funny behaviour in the survey because of that `onBlur` change in the `CustomTextField` component. Long story short: for the survey, when you hit save, if there are errors, we have some code that scrolls the page to that component, and this is getting tripped up by that onBlur change (its scrolling randomly in the middle of typing into fields, etc). Now, that in mind, the scrolling behaviour was never quite perfect and needed some refactoring, which we never got to. So quite possibly we could remove it, otherwise we'd have to make that onBlur prop optional, etc. Probably a question for @jeznorth when hes back on how he wants the form to operate, so can maybe ignore it for now.

File: app/src/components/fields/CbSelectField.tsx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+/**
+ *
+ * Wrapper for formik selects to handle all errors / onChange / onBlur
+ *
+ * Returns {*}
+ */
+
+export const FormikSelectWrapper = ({
```
Comment (NickPhura) [inline app/src/components/fields/CbSelectField.tsx]: If this is a generic component, can we move it into its one file?

File: api/src/app.ts (hunk 1 above)
Comment (MacDeluca) [reply api/src/app.ts]: Fixed

File: app/src/components/dialog/EditDialog.tsx (hunk 2 above)
Comment (MacDeluca) [reply app/src/components/dialog/EditDialog.tsx:26]: Fixed this and updated other components to use correct JSDoc formatting

File: app/src/contexts/configContext.tsx (hunk 3 above)
Comment (MacDeluca) [reply app/src/contexts/configContext.tsx]: Fixed

File: app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx (hunk 4 above)
Comment (MacDeluca) [reply app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx]: Fixed

File: app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx (hunk 5 above)
Comment (MacDeluca) [reply app/src/features/surveys/view/survey-animals/form-sections/MeasurementAnimalForm.tsx]: Removed the key here. Parent rendering list of components above those this isnt needed.

File: app/src/components/fields/CbSelectField.tsx (hunk 6 above)
Comment (MacDeluca) [reply app/src/components/fields/CbSelectField.tsx]: Moved into fields directory / renamed to CbSelectFieldWrapper

**Comment by MacDeluca** [comment]
"Getting some warnings in the form for a bunch of the array sections.
Encountered two children with the same key, '[object Object]'."

Where did you see these errors? I checked console and the app logs

**Comment by NickPhura** [comment]
@MacQSL 
> "Getting some warnings in the form for a bunch of the array sections. Encountered two children with the same key, '[object Object]'."
> 
> Where did you see these errors? I checked console and the app logs

It shows up when you add more than 1 marking (I think the same happens for the other array sections too, but I didn't explicitly check)

image

**Comment by NickPhura** [comment]
@MacQSL 
Just testing the UI more closely and noticing a few things:

### Capture/Mortality Event
- If I add a comment, and then toggle to the map tab and back, the comment field gets collapsed (the value is retained, if I re-open it).
- Capture coordinate uncertainty defaults to `10`, but release coordinate uncertainty defaults to empty.
  - Looking at the Post body, capture uncertainty is sent as a number, but release uncertainty is sent as a string (if it matters).
- Mortality uncertainty defaults to `1` (should it also be 10?)

### Markings (Caribou in this case)
- For a single marking entry, If I pick different measurement types back and forth, the value field gets into a bad state. Below screenshot shows 2 markings I was messing with, and how they show different value fields:
  - image

### Also getting a key error when initially opening the animal dialog
- image

**Comment by MacDeluca** [comment]
@NickPhura Resolved the issues in the above comment

**Comment by al-rosenthal** [comment]
I tested this and noticed a few things:

SIMS is now on mui 5 so it would be great if the date pickers were updated to use the mui date picker. An example of these is in the project create form

When Moose, Grizzly or any of the Animal Families (Alces, Canidae, Ursidae ect) are selected under the Markings form, the `Marking Type` field is still works but the `Marking Body Location` is disable but still required by the form. If it isn't required for those animals it shouldn't block the form. 

With Animal Measurements if `Estimated Age` is selected the value field specifies `Value [kilogram's]`

nit pick: In a capture event, if I toggle the UTM Coordinates it fills the coordinates with the center point of the map. Not sure if this is a bug or not just something I noticed

**Comment by MacDeluca** [comment]
@al-rosenthal 
- Good suggestion on the MUI date pickers i'll update those now

- I'll update the marking section to prevent it from being blocked if no values exist.

"With Animal Measurements if Estimated Age is selected the value field specifies Value [kilogram's]"
- Yea this part is kinda silly, its just referencing the measurement unit in the database. I set it to something temporarily [kilograms] in Critterbase dev for testing. Not an issue in production.

- I think the utm corrdinate is just generating off of 0,0 this can probably be removed and leave black till they fill it in.

Update: All issues mentioned resolved.

**Review by al-rosenthal — APPROVED** [review]
🌮

**Review by NickPhura — APPROVED** [review]

**Pull request merged**